Fold points directly into the `Id`

ullingerc opened the pull request:
Points (specified by WKT literals with type point) are now stored directly in the `ValueId`, with 30 bits of precision for each of the two coordinates.  This makes working on point geometries (e.g. spatial queries that only work on centroids) orders of magnitude faster, as we save the retrieving of the points from disk as well as parsing them.

joka921 reviewed (changes requested):
Thanks a lot, that was very quick. 
We have already tried it out, and it works like a charm and is much much faster.
I have left you some comments and will send you an email with detailed instructions.
Keep up the good work!

ullingerc commented on src/global/GeoPoint.h:
```diff
@@ -0,0 +1,69 @@
+
+  std::string toStringRepresentation() {
+    return absl::StrCat("POINT(", std::to_string(this->getLng()), " ",
+                        std::to_string(this->getLat()), ")");
```
StrCat automatically converts doubles to strings, however with less decimal places than `std::to_string`. That's why I used the latter. 

For example with `std::to_string` I get:
`Restaurant InterCity Hotel	POINT(7.841799 47.997157)`

But without I get:
`Restaurant InterCity Hotel	POINT(7.8418 47.9972)`

ullingerc commented on test/SparqlExpressionTest.cpp:
```diff
@@ -1105,6 +1105,7 @@ TEST(SparqlExpression, testToNumericExpression) {
 
 // ____________________________________________________________________________
 TEST(SparqlExpression, geoSparqlExpressions) {
+  // TODO<ullingerc>: Fix tests after implementation of GeoPoint
```
I have updated the tests and also added many more tests in GeoPointTest.cpp. However there are still some issues with them.

joka921 commented on test/SparqlExpressionTest.cpp:
```diff
@@ -1109,20 +1109,10 @@ TEST(SparqlExpression, geoSparqlExpressions) {
-            IdOrLiteralOrIri{lit("NotAPoint")});
-  checkDist(U, IdOrLiteralOrIri{lit("NotAPoint")},
-            IdOrLiteralOrIri{lit("POINT(24.3 26.8)")});
+  auto v = ValueId::makeFromGeoPoint(GeoPoint(26.8, 24.3));
```
```
autot actualLat = Point(fromBits(point.toBits)).lat_;
EXPECT_FLOAT_NEAR(actualLat, 26.8, somePrecision);
// build a DoubleId from the actualLat and pass that to the `checkLat`function as expected
```
Same for long.
add a comment that the toBit and fromBit has some precision loss as there are only 30 bits per coordinate.

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -0,0 +1,73 @@
+    auto [lng, lat] = ad_utility::detail::parseWktPoint(
+        asStringViewUnsafe(value.getContent()));
+    if (!std::isnan(lng) && !std::isnan(lat)) {
+      return GeoPoint(lat, lng);
```
```suggestion
      return GeoPoint{lat, lng};
```

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -0,0 +1,73 @@
+std::optional<GeoPoint> GeoPoint::parseFromLiteral(
+    const ad_utility::triple_component::Literal& value) {
+  if (value.hasDatatype() &&
+      value.getDatatype().compare(asNormalizedStringViewUnsafe(
+          std::string_view(GEO_WKT_LITERAL))) == 0) {
```
```suggestion
      value.getDatatype() == asNormalizedStringViewUnsafe(
          std::string_view(GEO_WKT_LITERAL))) {
```
Should work.
Probably the `std::string_view()` is also redundant.

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -0,0 +1,73 @@
+  double lng =
+      extractCoordinate(bits, coordinateMaskLng, 0, COORDINATE_LNG_MAX);
+
+  return GeoPoint(lat, lng);
```
```suggestion
  return {lat, lng};
```

joka921 replied:
Only works for non explicit constructor.s

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -0,0 +1,94 @@
+#include "parser/Literal.h"
+#include "util/SourceLocation.h"
+
+class Literal;
```
```suggestion
```

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -0,0 +1,94 @@
+      (double)(1 << numDataBitsCoordinate);
+
+  // Construct GeoPoint and ensure valid coordinate values
+  GeoPoint(double lat, double lng) {
```
```suggestion
  GeoPoint(double lat, double lng) : lat_{lat}, lng_{lng} {
```

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -0,0 +1,94 @@
+    if (lng < -COORDINATE_LNG_MAX || lng > COORDINATE_LNG_MAX)
+      throw CoordinateOutOfRangeException(lng, false);
+
+    lat_ = lat;
+    lng_ = lng;
```
```suggestion
```

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -0,0 +1,94 @@
+
+  // Convert the value of this GeoPoint object to a single bitstring.
+  // The conversion will reduce the precision and thus change the value.
+  // However the lost precision should only be in the range of centimeters.
```
Also document that the first (how many) bytes are guaranteed to be 0 , that is needed by the ValueId.

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -0,0 +1,94 @@
+
+  std::pair<std::string, const char*> toStringAndType() {
+    return std::pair(toStringRepresentation(), GEO_WKT_LITERAL);
+  }
```
Probably all the functions (toStringRep, toStringAndType, probably also the constructor) can be in the `.cpp`

joka921 commented on src/parser/TripleComponent.cpp:
```diff
@@ -49,18 +49,8 @@ std::optional<Id> TripleComponent::toValueIdIfNotString() const {
-        }
-      }
+      auto geopoint = GeoPoint::parseFromLiteral(value);
+      if (geopoint.has_value()) return Id::makeFromGeoPoint(geopoint.value());
```
```suggestion
      if (geopoint.has_value()) {return Id::makeFromGeoPoint(geopoint.value()); }
```

joka921 commented on src/util/GeoSparqlHelpers.cpp:
```diff
@@ -31,33 +38,17 @@ std::pair<double, double> parseWktPoint(const std::string_view point) {
     // coordinate strings should also parse to a double.
-    AD_CONTRACT_CHECK(lng != invalidCoordinate);
-    AD_CONTRACT_CHECK(lat != invalidCoordinate);
+    AD_CONTRACT_CHECK(!std::isnan(lng));
+    AD_CONTRACT_CHECK(!std::isnan(lat));
```
```suggestion
    AD_CORRECTNESS_CHECK(!std::isnan(lng));
    AD_CORRECTNESS_CHECK(!std::isnan(lat));
```

joka921 commented on src/util/GeoSparqlHelpers.cpp:
```diff
@@ -31,33 +38,17 @@ std::pair<double, double> parseWktPoint(const std::string_view point) {
-  }
-  return lng;
-}
+double wktLongitudeImpl(GeoPoint point) { return point.getLng(); }
```
Probably those can now be inlined directly in the class in the header.

joka921 commented on src/util/GeoSparqlHelpers.cpp:
```diff
@@ -31,33 +38,17 @@ std::pair<double, double> parseWktPoint(const std::string_view point) {
-  return lat;
-}
+// Parse latitude from WKT point.
+double wktLatitudeImpl(GeoPoint point) { return point.getLat(); }
```
Same here.

joka921 commented on src/util/GeoSparqlHelpers.h:
```diff
@@ -32,48 +25,47 @@ static constexpr double invalidCoordinate =
+    if (!point.has_value()) {
+      return std::numeric_limits<double>::quiet_NaN();
+    }
+    return detail::wktLatitudeImpl(point.value());
```
No need for the `Impl`, write `point.value().getLat` direclty here, and clean the rest up.

joka921 commented on test/GeoPointTest.cpp:
```diff
@@ -2,16 +2,121 @@
+  constexpr double lng = ((-130.2 + 180) / (2 * 180)) * (1 << 30);
+  ASSERT_EQ(g.toBitRepresentation() & ((1 << 30) - 1), round(lng));
+
+  constexpr size_t expect = (static_cast<size_t>(1) << 30) << 30 | (1 << 30);
```
This leaks into the datatype bits, fix it, and add an assertion in the to"BitRepresentation, that those are in fact left untouched.

joka921 commented on test/SparqlExpressionTest.cpp:
```diff
@@ -1105,25 +1105,14 @@ TEST(SparqlExpression, testToNumericExpression) {
-  checkDist(U, IdOrLiteralOrIri{lit("NotAPoint")},
-            IdOrLiteralOrIri{lit("POINT(24.3 26.8)")});
+  auto v = ValueId::makeFromGeoPoint(GeoPoint(26.8, 24.3));
+  checkLat(v, Ids{D(26.8)});
```
Keep the tests above, just replace the valid point literals by actual geoPoints .

joka921 reviewed (changes requested):
A round of reviews.

ullingerc commented:
@joka921 Thank you very much. I have implemented all your reviews and updated the tests. They now all pass. From my part, this PR should now be ready.

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -49,19 +46,13 @@ class GeoPoint {
+      ~(coordinateMaskLat | coordinateMaskLng);
   static constexpr double maxCoordinateEncoded =
-      (double)(1 << numDataBitsCoordinate);
+      (double)((1 << numDataBitsCoordinate) - 1);
```
For better readability, can you use the helpers in `BitUtils.h` for these masks, so e.g.
`coordinateMaskLng = bitMaskForLowerBits(numDataBitsCoordinate)`
`coordinateMaskFreeBits = bitMaskForHigherBits(sizeof(T) * 8 - numDataBits)`
`maxCoordinateEncoded = static_cast<double>(coordinateMaskLat)`

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -70,25 +61,22 @@ class GeoPoint {
   // The conversion will reduce the precision and thus change the value.
   // However the lost precision should only be in the range of centimeters.
-  T toBitRepresentation();
+  // Guarantees to only use the lower numDataBits (currently 60 bits),
```
```suggestion
  // Guarantees to only use the lower `numDataBits` (currently 60 bits),
```
(we use backticks for variable and type names to distinguish them from "ordinary" words)

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -70,25 +61,22 @@ class GeoPoint {
   // Restore a GeoPoint object from a single bitstring produced by the above
   // function. Due to the reduction of precision this object will not have the
-  // identical value.
+  // identical value. Ignores the upper 4 bits (only uses the lower numDataBits)
```
```suggestion
  // identical value. Ignores the upper 4 bits (only uses the lower `numDataBits`)
```
(see above)

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -69,5 +92,15 @@ GeoPoint GeoPoint::fromBitRepresentation(T bits) {
-  return GeoPoint(lat, lng);
+  return {lat, lng};
+};
+
```
```suggestion
// ______________________________________________________________________________
```

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -69,5 +92,15 @@ GeoPoint GeoPoint::fromBitRepresentation(T bits) {
+  return absl::StrCat("POINT(", std::to_string(getLng()), " ",
+                      std::to_string(getLat()), ")");
+};
+
```
Same here.

joka921 commented on src/parser/TripleComponent.cpp:
```diff
@@ -50,7 +50,9 @@ std::optional<Id> TripleComponent::toValueIdIfNotString() const {
     } else if constexpr (std::is_same_v<T, Literal>) {
       auto geopoint = GeoPoint::parseFromLiteral(value);
-      if (geopoint.has_value()) return Id::makeFromGeoPoint(geopoint.value());
+      if (geopoint.has_value()) {
+        return Id::makeFromGeoPoint(geopoint.value());
+      }
```
I know that I initially told you to put this implementation here,
but now I remember that this was just to initially get something working.
You should move this logic to
the function `literalAndDatatypeToTripleComponentImpl` in the `TurtleParser` class in the `RdfParser.cpp` file (line 460ff).

There we have all the other logic for folding types etc. Then we have one centralized place, that ensures that we don't miss anything anywhere.

joka921 reviewed (changes requested):
This is now much much cleaner and seems to work nicely.
I have some final requests.

Additionally your new datatype in the ValueId has left some gaps in the coverage (check the CodeCOV report). You can fix them by adding tests e.g. to some other sparql expressions, s.t. they also have test cases with GeoPoints. and in the ValueIdTest.h you can test the name etc.

joka921 reviewed (changes requested):
This also is an index-breaking change, so you have to change the values at the end of 
`IndexFormatVersion.h` (use the number of your PR, the date we can change later).

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -60,9 +60,10 @@ GeoPoint::T GeoPoint::toBitRepresentation() const {
+    const ad_utility::triple_component::Literal& value, bool checkDatatype) {
+  if ((value.hasDatatype() &&
+       value.getDatatype() == asNormalizedStringViewUnsafe(GEO_WKT_LITERAL)) ||
+      !checkDatatype) {
```
If you want this as a micro-optimization, then you should put the `!checkDatatype` first, to make it short-circuit

joka921 commented on src/parser/GeoPoint.cpp:
```diff
@@ -95,12 +96,19 @@ GeoPoint GeoPoint::fromBitRepresentation(T bits) {
 };
+
+// _____________________________________________________________________________
+bool GeoPoint::operator==(const GeoPoint& other) const {
```
As these are the only members, you can default this operator, but the following has to be in the header:

`bool operator==(const GeoPoint& other) const = default`
(Very handy C++20 feature).

joka921 commented on src/parser/GeoPoint.h:
```diff
@@ -61,22 +68,26 @@ class GeoPoint {
 
   std::pair<std::string, const char*> toStringAndType() const;
+
+  bool operator==(const GeoPoint& other) const;
```
```suggestion
  bool operator==(const GeoPoint& other) const = default;
```
see comment above.

joka921 commented on test/RdfParserTest.cpp:
```diff
@@ -253,6 +256,10 @@ TEST(RdfParserTest, literalAndDatatypeToTripleComponent) {
+  auto result = ladttc("POINT(7.8 47.9)", fromIri(GEO_WKT_LITERAL));
+  auto vid = result.toValueIdIfNotString();
+  ASSERT_TRUE(vid.has_value() &&
+              vid.value().getDatatype() == Datatype::GeoPoint);
```
I think for the test coverage you also need a WKT-Literal that is not a point to test that `else` branch.

joka921 reviewed (changes requested):
Great, only very very minor changes remaining.

joka921 reviewed (approved):
Thank you very much, 
I will wait for (at least the not very slow) checks to run through, and then we can merge this.
